[release/10.0.1xx] Package Pruning data fixes for past frameworks

ericstj opened the pull request:
Fixes https://github.com/dotnet/sdk/issues/51320

## Customer Impact

- [x] Customer reported
- [x] Found internally

1. Referencing System.Collections.Immutable, System.ComponentModel.Composition, System.Reflection.TypeExtensions, System.Runtime.Loader, System.Security.AccessControl, System.Security.Principal.Windows from a NETStandard2.1 build would fail to compile after enabling pruning.
2. System.DirectoryServices was pruned at too high a version, which could lead to ref-def mismatch errors at runtime in frameworks older than `net8.0`, as the framework copy would be used instead of package.  _(unsupported framework versions)_
3. All of WindowsDekstop in `net9.0` was pruning only 8.0 package versions.  No customer impact other than pruning less than we could.
4. System.Drawing.Common was pruned at too high a version on `netcoreapp3.0` which could lead to ref-def mismatch errors at runtime.  _(unsupported framework version)_
5. Many packages in `netstandard2.0` were pruned at lower versions due to inconsistent facade assembly versioning in netstandard (corefx servicing interaction with standard repo).   Many of these packages would be pruned by other packages which depended on them, but not directly.  No customer impact other than pruning less than we could.
6. Referencing System.Security.Cryptography.Pkcs from ASPNETCore would fail to compile after enabling pruning.  This package was not exposed for reference, so it shouldn't be pruned. https://github.com/dotnet/aspnetcore/issues/64125
  
## Regression

- [x] Yes
- [ ] No

Regression caused when enabling pruning.  Customer with a project that works will face these issues when adding a `net10` target to a multi-targeted project, or manually enabling package pruning.

## Testing

Fixed the initial customer report, then did extensive adhoc testing by restoring pruned package versions for every framework version and comparing the result to references exposed for that framework.

https://github.com/ericstj/scratch/tree/pruningTest

Example tests:
1. Assemblies missing for framework but available from pruned packages.
2. Assembly version mismatch between pruned (higher) and framework.
3. API Compatibility issue between pruned package and framework.
4. Later version of pruned framework prunes less than previous.

## Risk

Low: In most cases pruned data is pruning less.  Test methodology can observe the results of these changes to ensure they don't have side-effects other than desired.

@dsplaisted @nkolev92

ericstj committed:
Remove pruned packages from netstandard2.1 which were listed in packageoverrides but not present in references

ericstj commented on src/Tasks/Microsoft.NET.Build.Tasks/FrameworkPackages/FrameworkPackages.netstandard2.1.cs:
```diff
@@ -27,20 +26,13 @@ internal static class NETStandard21
             { "System.Net.WebHeaderCollection", "4.3.0" },
             { "System.Numerics.Vectors", "4.6.1" },
             { "System.ObjectModel", "4.3.0" },
-            { "System.Private.DataContractSerialization", "4.3.0" },
```
People won't reference this and all public paths to that package will be pruned.  We don't have to make the pruning data closure-complete.  There is no "customer facing" behavior that changes as a result of that.

> theoretically be a problem.

Please describe the customer-facing problem you imagine.

nkolev92 replied:
They won't, but they still can no? 

I think making the data closure complete makes it easy to establish a standard for pruning which would you can only reference prunable packages. 

That being said, with ns2.0, once we get it right, it doesn't matter. 
It's the net10.0 and newer data that'll keep evolving, so we could enforce that standard there.

nkolev92 replied:
I agree with Eric's reasoning. We had a chat offline as well.

ericstj committed:
Remove System.ComponentModel.Composition from ns2.1 pruning

This assembly is present, but its a minimal compat facade.  It should not take the place of the package.

ericstj committed:
Fix pruning data for System.DirectoryServices

This assembly does not version in the same way as others since it's 1/1 with it's .NETFramework name.

The tooling which calculated these pruning lists didn't catch that.

Also updating the net9 WindowsDesktop pruning data to capture the GA versions.

ericstj committed:
Fix incorrect Drawing.Common reference on .NETCore 3.0

ericstj committed:
Raise package versions in netstandard2.0 to be consistent

ericstj committed:
Remove System.Security.Cryptography.Pkcs from ASPNetCore pruning data

This assembly is part of ASPNETCore, but never exposed from the targeting pack

jeffschwMSFT reviewed (approved):
approved
